### ISSUE TITLE
Mixtral QLoRA BF16 error

### ISSUE THREAD

**Issue by generalsvr** [axolotl-ai-cloud/axolotl#938]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Expected to run fine-tuning task

### Current behaviour

```
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/utils/operations.py", line 659, in forward
    return model_forward(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/utils/operations.py", line 647, in __call__
    return convert_to_fp32(self.model_forward(*args, **kwargs))
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/amp/autocast_mode.py", line 14, in decorate_autocast
    return func(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/peft/peft_model.py", line 977, in forward
    return self.base_model(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/peft/tuners/tuners_utils.py", line 106, in forward
    return self.model.forward(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 164, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/models/mixtral/modeling_moe_mistral.py", line 1259, in forward
    outputs = self.model(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 164, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/models/mixtral/modeling_moe_mistral.py", line 1109, in forward
    layer_outputs = self._gradient_checkpointing_func(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 249, in checkpoint
    return CheckpointFunction.apply(function, preserve, *args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 107, in forward
    outputs = run_function(*args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 164, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/models/mixtral/modeling_moe_mistral.py", line 808, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 164, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/models/mixtral/modeling_moe_mistral.py", line 536, in forward
    attn_output = flash_attn_varlen_qkvpacked_func(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/flash_attn/flash_attn_interface.py", line 752, in flash_attn_varlen_qkvpacked_func
    return FlashAttnVarlenQKVPackedFunc.apply(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/flash_attn/flash_attn_interface.py", line 255, in forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = _flash_attn_varlen_forward(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/flash_attn/flash_attn_interface.py", line 79, in _flash_attn_varlen_forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = flash_attn_cuda.varlen_fwd(
RuntimeError: FlashAttention only support fp16 and bf16 data type
```

### Steps to reproduce

Run official mixtral config on Runpod A100. 

Docker winglian/axolotl-runpod:main-latest

transformers @ git+https://github.com/huggingface/transformers.git@df5c5c62ae253055336f5bb0828ca8e3e15ab6bd

### Config yaml

```
base_model: DiscoResearch/mixtral-7b-8expert
model_type: MixtralForCausalLM
tokenizer_type: LlamaTokenizer
trust_remote_code: true

load_in_8bit: false
load_in_4bit: true
strict: false

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca
dataset_prepared_path: last_run_prepared
val_set_size: 0.0
output_dir: ./qlora-out

adapter: qlora
lora_model_dir:

sequence_len: 4096
sample_packing: true
pad_to_sequence_len: true

lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
lora_target_linear: true
lora_fan_in_fan_out:
#lora_target_modules:
#  - gate
#  - q_proj
#  - k_proj
#  - v_proj
#  - o_proj
#  - w1
#  - w2
#  - w3

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 2
micro_batch_size: 1
num_epochs: 1
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.0002

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

loss_watchdog_threshold: 5.0
loss_watchdog_patience: 3

warmup_steps: 10
eval_steps:
eval_table_size:
eval_table_max_new_tokens: 128
save_steps:
debug:
deepspeed: deepspeed/zero2.json
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10.13

### axolotl branch-commit

main/86487c2

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by JustinMeans** [comment]
I'm also getting this error with on a newly installed RunPod Docker Image for A100, with the same config yaml and transformers version referenced above. 

```
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules
/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
[2023-12-12 01:51:15,523] [INFO] [axolotl.normalize_config:141] [PID:2960] [RANK:0
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules
/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks
[2023-12-12 01:52:35,057] [INFO] [axolotl.normalize_config:141] [PID:3042] [RANK:0
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules
/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks
.py", line 164, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/models/mixtral/modeling_moe_mistral.py", li
ne 808, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules
/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks
.py", line 164, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/models/mixtral/modeling_moe_mistral.py", li
ne 536, in forward
    attn_output = flash_attn_varlen_qkvpacked_func(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/flash_attn/flash
_attn_interface.py", line 752, in flash_attn_varlen_qkvpacked_func
    return FlashAttnVarlenQKVPackedFunc.apply(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/autograd/f
unction.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/flash_attn/flash
_attn_interface.py", line 255, in forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = _flash_attn_varlen
_forward(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/flash_attn/flash
_attn_interface.py", line 79, in _flash_attn_varlen_forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = flash_attn_cuda.va
rlen_fwd(
RuntimeError: FlashAttention only support fp16 and bf16 data type
```

**Comment by Undi95** [comment]
Same here, tried on A100, L40 and A6000 on runpod.

This fixed the issue : https://github.com/OpenAccess-AI-Collective/axolotl/pull/802/files
![image](https://github.com/OpenAccess-AI-Collective/axolotl/assets/35925808/4b022823-fdd9-4d8e-8d90-9fe01d1f3f60)
![image](https://github.com/OpenAccess-AI-Collective/axolotl/assets/35925808/7d34ca6d-80f0-4221-9521-01575b996afa)

**Comment by JustinMeans** [comment]
This issue was fixed by PR 942 for me: https://github.com/OpenAccess-AI-Collective/axolotl/pull/942

Afterwards I faced the DummyOptim issue which this addresses (remove DeepSpeed in config yaml for single GPU):
https://github.com/OpenAccess-AI-Collective/axolotl/pull/802/files

Now fine-tuning is underway 🚀
<img width="810" alt="Screenshot 2023-12-12 at 12 00 12 AM" src="https://github.com/OpenAccess-AI-Collective/axolotl/assets/46542161/86e6b7e9-d146-40e2-a5ed-f48e34ef5087">

**Comment by hamelsmu** [comment]
This issue was closed by #942

**Comment by dong-river** [comment]
Hi @generalsvr, I wonder if you have successfully fine-tuned Mixtral. If so, would you mind sharing the hyperparameter configuration? In particular, I am curious what vRAM, batch size, and learning rate I would expect
